docs/proposed/conflict-api.rst

```diff
@@ -11,42 +11,77 @@ This document is aimed at programmers working on magic-folder. It is a proposed
 =====================================
 
 
-
-``GET /v1/conflicts/<folder-name>``
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+``GET /v1/magic-folder/<folder-name>/conflicts``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Returns a list (possibly empty) of local filesystem paths corresponding to each Snapshot that is currently in a Conflict state in the given magic-folder.
 
 Our content is in the path itself.
-The conflicting "other" content is in ``<path>.conflict-<name>`` where ``<name>`` is the petname of the user who is provided the conflicted content.
+The conflicting "other" content is in ``<path>.conflict-<name>`` where ``<name>`` is the petname of the participant who is provided the conflicted content.
+
+
 
-Justification: we need somewhere for "theirs" versus "my" content .. I think we should still reflect this on the filesystem, even if the *API to manipulate it* is no longer there.
-This makes it more obvious for CLI users that they should check the conflicts list; the only alternative would seem to be "run some command occasionally to check for conflicts".
-I personally will forget to run this command.
+``POST /v1/magic-folder/<folder-name>/resolve-conflict
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+A JSON body is passed to this endpoint, following this example::
 
-``POST /v1/resolve_conflict/<folder-name>?path=<some-path>&resolution=<theirs|mine>``
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    {
+        "relpath": "local/path",
+        "resolution": "author"
+    }
 
-The ``path`` query argument is required.
+The ``relpath`` key is required.
 It must be a filesystem path relative to the selected magic-folder.
 
-The ``resolution`` query argument is required.
-It must be either the string ``theirs`` or the string ``mine``.
+The ``resolution`` key is required.
+It must be the name of a current participant in the given magic-folder.
 
-It is an error if the given ``path`` in the given magic-folder is not currently in a conflicted state. In this case the response code is **404 Not Found** (XXX is this appropriate?)
+It is an error if the given ``relpath`` in the given magic-folder is not currently in a conflicted state.
 
-If the resolution is ``theirs`` then the file at ``<path>.theirs.<name>`` is moved to ``<path>`` and a new (local) Snapshot is created (with two parents).
+If the ``resolution`` is our user-name then all conflict files are deleted new (local) Snapshot is created (with parents corresponding to all conflicting participants).
 
-If instead the resolution is ``mine`` then the file at ``<path>.theirs.<name>`` is deleted and a new (local) Snapshot is created (with two parents).
+If instead the resolution is some other participant, then the content of ``<relpath>.conflict-<participant>`` is moved to ``<relpath>`` and any other conflict files are deleted.
+Then a new (local) Snapshot is created (with parents corresponding to all conflicting participants).
```

Comment:
I'm not sure exactly how this is intended to be used, but it seems there is somewhat of a race condition here, we could have detected a conflict with a new device between when the user indicated to their UI that they had resolved all the conflicts, and when we get notified. (Or if conflict files are themselves snapshotted, one of them could have been updated in that interval.

I think we may want to pass the participants in the conflict we intend to resolve here, rather than assuming all of them (or some otherway of indicating which state is being resolved).

Reply:
We stop downloading once a conflict is detected so I don't think that can happen.

(Anyway, we don't implement this part yet -- there's no way out of a conflict yet, so this can be updated further when that is implemented)